SourceX/DiabloUI/ui_item.h

```diff
@@ -38,316 +38,321 @@ enum UiFlags {
 	}
 
-	char *value;
-	std::size_t max_length;
-};
+	enum FrameKey {
+		DEFAULT = 0,
+		PRESSED,
+		DISABLED
+	};
 
-struct UiItem {
-	constexpr UiItem(UiText text)
-	    : type(UI_TEXT)
-	    , text(text)
-	{
-	}
+	//private:
+	Art *m_art;
 
-	constexpr UiItem(UiArtText text)
-	    : type(UI_ART_TEXT)
-	    , art_text(text)
-	{
-	}
+	const char *m_text;
+	void (*m_action)();
 
-	constexpr UiItem(UiArtTextButton art_text_button)
-	    : type(UI_ART_TEXT_BUTTON)
-	    , art_text_button(art_text_button)
-	{
-	}
+	// State
+	bool m_pressed;
+	TtfSurfaceCache *m_render_cache;
 
-	constexpr UiItem(UiImage image)
-	    : type(UI_IMAGE)
-	    , image(image)
+	virtual void FreeCache()
 	{
+		delete m_render_cache;
+		m_render_cache = NULL;
 	}
+};
 
-	constexpr UiItem(UiButton button)
-	    : type(UI_BUTTON)
-	    , button(button)
-	{
-	}
+//=============================================================================
 
-	constexpr UiItem(UiList list)
-	    : type(UI_LIST)
-	    , list(list)
+class UiListItem {
+public:
+	UiListItem(const char *text = "", int value = 0)
 	{
+		m_text = text;
+		m_value = value;
 	}
 
-	constexpr UiItem(UiScrollBar scrollbar)
-	    : type(UI_SCROLLBAR)
-	    , scrollbar(scrollbar)
+	~UiListItem()
 	{
 	}
 
-	constexpr UiItem(UiEdit edit)
-	    : type(UI_EDIT)
-	    , edit(edit)
-	{
-	}
+	//private:
+	const char *m_text;
+	int m_value;
+};
 
-	UiType type;
-	union {
-		UiText text;
-		UiArtText art_text;
-		UiImage image;
-		UiArtTextButton art_text_button;
-		UiButton button;
-		UiList list;
-		UiScrollBar scrollbar;
-		UiEdit edit;
-		UiItemBase common;
-	};
+typedef std::vector<UiListItem *> vUiListItem;
 
-	bool has_flag(UiFlags flag) const
+class UiList : public UiItemBase {
+public:
+	UiList(vUiListItem vItems, Sint16 x, Sint16 y, Uint16 item_width, Uint16 item_height, int flags = 0)
+	    : UiItemBase(x, y, item_width, item_height * vItems.size(), flags)
 	{
-		return common.has_flag(flag);
-	}
+		m_type = UI_LIST;
+		m_vecItems = vItems;
+		m_x = x;
+		m_y = y;
+		m_width = item_width;
+		m_height = item_height;
+	};
 
-	bool has_any_flag(int flags) const
+	~UiList() {};
+
+	SDL_Rect itemRect(int i)
 	{
-		return common.has_any_flag(flags);
+		SDL_Rect tmp;
+		tmp.x = m_x;
+		tmp.y = m_y + m_height * i;
+		tmp.w = m_width;
+		tmp.h = m_height;
+
+		return tmp;
 	}
 
-	const SDL_Rect &rect() const
+	UiListItem *itemAt(Sint16 y) const
 	{
-		return common.rect;
+		ASSERT(y >= m_rect.y);
+		const std::size_t index = (y - m_rect.y) / m_height;
+		ASSERT(index < m_vecItems.size());
+		return m_vecItems[index];
 	}
 
-	void FreeCache()
+	UiListItem *GetItem(int i)
 	{
-		switch (type) {
-		case UI_BUTTON:
-			button.FreeCache();
-			break;
-		case UI_TEXT:
-			text.FreeCache();
-			break;
-		default:
-			break;
-		}
+		return m_vecItems[i];
 	}
+
+	//private:
+	Sint16 m_x, m_y;
+	Uint16 m_width, m_height;
+	std::vector<UiListItem *> m_vecItems;
 };
 
+typedef std::vector<UiItemBase *> vUiItemBase;

```

Comment:
Typedefing `vector` makes the code harder to read. It also makes it unobvious when something expensive happens, such as an entire vector is being copied (e.g. passed-by-value). Or when `delete` isn't done on each element.